### ISSUE TITLE
User orders with pre- and post-interactions

### ISSUE THREAD

**Issue by MartinquaXD** [cowprotocol/services#1457]
### Requirements
1. User needs to be able to tell us about the pre/post interactions (API)
2. We need to verify that they work (simulation)
3. We need to store them (DB)
4. We need to recover them (DB)
5. We need to price them into the fee (simulation, decide the behavior in relation to different order types)
6. We need to encode them (driver, add interactions to auction)
7. Gnosis solvers need to have support for this feature (probably a no-op if we disallow partially fillable orders)

### Possible Problems
- Some orders can have mutually exclusive pre/post interactions. How to deal with those?
- It's unclear how this feature should work for partially fillable orders.
    Should we execute the pre interactions only the first time we touch a partially fillable order?
    Should we execute the post interactions only when the order was completely filled?
    How would we inform solvers about the additional gas cost so they can incorporate that in their score? (this assumes that we don't want to make solvers simulate all the custom interactions to compute the additional fee on their own)
    I think this feature should probably be disabled for partially fillable orders.
- Who is responsible for uploading the IPFS appdata? (usually `/quote` returns something the user can easily sign, but that would require us to upload the IPFS data which seems wrong)

### Milestone 1 - "Quotes"
The `/quote` endpoint is able to accept custom interactions and we are able generate a quote for an order with custom interactions with correct fees.

#### Task 1 - Simulation helper
- we probably need to come up with some helper solidity code that:
    * simulates pre interactions
    * fakes the transfer of the actual trade
    * simulates the post interactions

#### Task 2 - Update POST `/quote` endpoint
- [x] #1487
- update openapi spec (mention the option to set custom interactions)
- simulate interactions to ensure they work and compute a fee (see task 1)
- probably some unforseen annoyance with storing the quote in the DB (TODO)

### Milestone 2 - "Order"
We can create an order with custom interactions and store them in the DB.

#### Task 1 - Proper support for reading `AppData`
- some CLI argument configuring the IPFS node
- decide on the format for custom interactions
- implement helpers to recover data from IPFS (helper should probably just return `serde_json::Value` and consumer can cast into correct format)

#### Task 2 - Update POST `/order` endpoint
- read `AppData` on request
- simulate custom interactions to ensure that order can be created
- [x] #1481 

### Milestone 3 - "Auction & Settlement"
We make the new orders part of the auction and can settle them.

#### Task 1 - Build auction with custom interactions
- ~~recover stored interactions from DB as part of the order metadata~~ (part of #1481)
- update openapi spec
- while building the auction simulate orders with custom interactions to see if they currently work (discard failing ones)
- make LimitOrderQuoter compatible with custom interactions

#### Task 2 - Encode tx
- [x] #1484

**Comment by nlordell** [comment]
Notes:

> I think this feature should probably be disabled for partially fillable orders.

This is a very good point that I did not think about. Lets continue working under that assumption for now.

> Who is responsible for uploading the IPFS appdata

This is a good question, and one that we should probably have a larger discussion around that involves @anxolin. Personally, I think it would be nice to completely abstract away the need to upload to IPFS from clients. So the FE would just send us an order with:

```js
// EITHER:
{
  // ...
  "appData": {
    // some valid app data object
  },
  // ...
}
// OR:
{
  // ...
  "appData": "0x...",
  "appDataDoc": {
    // some valid app data object, would error if `hash(appDataDoc) != appData`
  },
  // ...
}
```

The backend would compute the hash and use it for signature verification.

> Task 1 - Simulation helper [...] fakes the transfer of the actual trade

I think that it doesn't actually need to simulate the transfer but instead:
- Read the "available balance", i.e. `balanceOf(owner) - allowance(owner, vaultRelayer)`

I think simulating the transfer should instead be covered more generally by "trade simulation".

> Task 1 - Proper support for reading AppData

I think we should also store used AppData's in our database. Otherwise, we would need to "production-ize" our IPFS setup and _make sure_ that it is backed up and we can redeploy with all our AppData documents in case of Pinata failure.

---

I would also add:

### Milestone 4 - Refactor EthFlow

- Refactor EthFlow to use common pre- and post- interaction setup that we use for regular orders, thus reducing code paths.

**Comment by cowfee** [comment]
> We need to verify that they work (simulation)

Can this be done before solving? The post-interactions, in particular, would likely expect the settlement to have already executed. But I do think we should be able to simulate the pre-interactions in the autopilot.

**Comment by nlordell** [comment]
> Can this be done before solving?

We can verify that they execute in isolation. The limitation with that is that a cool post-interaction that relies on funds that are received as a result of a settlement (so relies on the transferred DAI amount in an order that buys DAI) might not work.

**Comment by MartinquaXD** [comment]
> I think that it doesn't actually need to simulate the transfer but instead:
> Read the "available balance", i.e. balanceOf(owner) - allowance(owner, vaultRelayer)
> I think simulating the transfer should instead be covered more generally by "trade simulation".

If I understood you correctly that means the post interactions are only possible if we do verified quotes. I think this is reasonable. We want to move to all verified quotes eventually. Just pointing that out so we are aware of it.

> I think it would be nice to completely abstract away the need to upload to IPFS from clients. So the FE would just send us an order with:

This sounds good. I was worried that the frontend and backend would have to be in sync for something like this but if the frontend just sends the object as they would construct it we should be good.
Also I didn't focus a lot on the frontend integration for this feature as it doesn't really seem all that useful. Would regular people actually have the need to craft some custom transactions for their trades?

> I think we should also store used AppData's in our database.

Makes sense to me. I was worried that `IPFS` and the objects we store in our DB would disagree but since we can verify that the contents hash to the expected value that should be okay.

> I would suggest that pre- and post- interactions always be executed through a trampoline that ignores reverts

Not sure about ignoring reverts. This gives the user way less guarantees. Seems like a tough problem to solve optimally. Will have to think more about it.

**Comment by nlordell** [comment]
> If I understood you correctly that means the post interactions are only possible if we do verified quotes. I think this is reasonable.

I don't think so - I think this is more important for `preInteractions`, but I don't think this is required per se. Specifically, I think my same argument of "the transfer in is verified by trade simulation" also applies to "the transfer in is verified by the bad token detector". One of the main reasons for simulating the `transfer` instead of `balanceOf` + `allowance` was to do fewer `eth_call`s.

> I was worried that the frontend and backend would have to be in sync for something like this but if the frontend just sends the object as they would construct it we should be good.

Yes, we would need to discuss with the FE before proceeding with this idea.

> Would regular people actually have the need to craft some custom transactions for their trades?

Not planned at all.

> This gives the user way less guarantees

This is a valid point. The way I see it, pre- and post- interactions are a "best effort" thing anyway, since there are no on-chain guarantees that order's interactions are are executing with these pre- and post- interactions anyway.

**Comment by anxolin** [comment]
> This is a good question, and one that we should probably have a larger discussion around that involves @anxolin. Personally, I think it would be nice to completely abstract away the need to upload to IPFS from clients. So the FE would just send us an order with:

Happy to have this conversation. Also seems reasonable to me, and probably best for other UIs (since uploading the doc is more involved than UIs would care about) --> i.e. we pre-calculate the hash, and in parallel upload and post the order. In case of error, we do re-attempts until it has uploaded. I don't think other UIs/CLIs etc will want to do this, so I think is best to provide the data with the order, so is also faster/easier for the backend to have this meta-info

Probably this discussion is better to have it outside of this issue, because is already quite long and has lots of other more important topics. 

By the way, is this issue's goal to have a generalised way to estimate on EIP1271 orders  and it means the responsibility of specifying the pre-interations is on the one asking the quote? 
Can you give some examples that might affect us if any? (like does it affect gasless safe orders, or any other)

**Comment by cowfee** [comment]
> Milestone 4 - Refactor EthFlow

Very good idea, and speaking of which: are we planning to keep this effort in parallel with trying to get colocation up and running in production?

**Comment by nlordell** [comment]
> Very good idea, and speaking of which: are we planning to keep this effort in parallel with trying to get colocation up and running in production?

I want to have co-location up and running. The split codebase is an annoying risk that we are carrying around and paying "technical debt interest" on.

**Comment by nlordell** [comment]
> it means the responsibility of specifying the pre-interations is on the one asking the quote?

Yes.

> Probably this discussion is better to have it outside of this issue

Good point, lets create a separate issue for tracking the "rich app-data" feature.

**Comment by vkgnosis** [comment]
> I would suggest that pre- and post- interactions always be executed through a trampoline that **ignores reverts**

This seems like a big deal. One use case of interactions is to uphold some guarantees. That isn't possible if interactions can't fail. I don't know what is correct but this feels really important and we need to make sure what use cases of interactions we want to support.

**Comment by vkgnosis** [comment]
> Who is responsible for uploading the IPFS appdata?

This part of the opening post isn't clear to me. It is the first time you mention appdata. Why is appdata involved in this issue? I think you mean that the pre and post interactions are stored in appdata but it isn't clear. The API could also take them as parameters and store in the database without involving appdata. Could we add motivation for this part?

**Comment by cowfee** [comment]
> > Who is responsible for uploading the IPFS appdata?
> 
> This part of the opening post isn't clear to me. It is the first time you mention appdata. Why is appdata involved in this issue? I think you mean that the pre and post interactions are stored in appdata but it isn't clear. The API could also take them as parameters and store in the database without involving appdata. Could we add motivation for this part?

I thought the appdata would be somehow passed as a parameter to the pre-interactions, but this isn't clear to me either.

Also, I need to get this point straight: are the pre- and post-interactions going to be signed? IMO it's really important that we sign them. For backwards compatibility, we could omit them from the signature if they are empty.

**Comment by nlordell** [comment]
> I thought the appdata would be somehow passed as a parameter to the pre-interactions, but this isn't clear to me either.

No, the pre- and post- interations are included in the data that generates the hash used in the `app_data` field.

> I need to get this point straight: are the pre- and post-interactions going to be signed?

Yes, basically:

```
let app_data = keccak256(something_including_pre_and_post_interactions);
let signature = sign_order(Order { app_data, ...rest_of_order_parameters });
```

So, we implicitly sign the pre- and post- interactions as part of an order (although, there is no on-chain verification for this).
